File: runtime/python/phonebook_rt/io_.py
"""Area 500 — the only block that touches the filesystem.

The CSV reader and writer are hand-written rather than delegated to the `csv`
module. That is deliberate: the Rust runtime implements the same explicit state
machine, so the two backends agree by construction instead of by hoping two
third-party libraries interpret the same edge cases the same way.
"""

from __future__ import annotations

from .faults import PhonebookFault

QUOTE_NEEDED = {",", '"', "\r", "\n"}


def read_text_file(path: str) -> str:
    """500-0000001 READ_TEXT_FILE — UTF-8 strict, no newline translation."""
    try:
        with open(path, "r", encoding="utf-8", errors="strict", newline="") as handle:
            return handle.read()
    except FileNotFoundError as exc:
        raise PhonebookFault("file_not_found", path) from exc
    except IsADirectoryError as exc:
        raise PhonebookFault("file_not_found", f"{path} is a directory") from exc
    except PermissionError as exc:
        raise PhonebookFault("permission_denied", path) from exc
    except UnicodeDecodeError as exc:
        raise PhonebookFault("decode_error", f"{path} is not valid UTF-8") from exc


def write_text_file(path: str, content: str) -> None:
    """500-0000002 WRITE_TEXT_FILE — truncating, byte-exact, no newline translation."""
    try:
        with open(path, "w", encoding="utf-8", newline="") as handle:
            handle.write(content)
    except PermissionError as exc:
        raise PhonebookFault("permission_denied", path) from exc
    except (FileNotFoundError, NotADirectoryError, IsADirectoryError) as exc:
        raise PhonebookFault("path_not_writable", path) from exc


def read_csv(path: str) -> list:
    """500-0000003 READ_CSV — header row, everything stays text."""
    text = read_text_file(path)
    records = _parse_csv(text)
    if not records:
        return []
    header = records[0]
    rows: list = []
    for line_no, cells in enumerate(records[1:], start=2):
        if len(cells) > len(header):
            raise PhonebookFault(
                "malformed_csv",
                f"{path}:{line_no} has {len(cells)} cells but the header has {len(header)}",
            )
        padded = list(cells) + [""] * (len(header) - len(cells))
        rows.append(dict(zip(header, padded)))
    return rows


def write_csv(path: str, rows: list, columns: list) -> None:
    """500-0000004 WRITE_CSV — minimal quoting, LF line endings."""
    lines = [_csv_line(columns)]
    for row in rows:
        lines.append(_csv_line([row.get(column, "") for column in columns]))
    write_text_file(path, "".join(line + "\n" for line in lines))


# --------------------------------------------------------------------------
# the shared CSV state machine
# --------------------------------------------------------------------------


def _parse_csv(text: str) -> list:
    records: list = []
    row: list = []
    field: list = []
    in_quotes = False
    i = 0
    size = len(text)

    while i < size:
        char = text[i]
        if in_quotes:
            if char == '"':
                if i + 1 < size and text[i + 1] == '"':
                    field.append('"')
                    i += 2
                    continue
                in_quotes = False
                i += 1
                continue
            field.append(char)
            i += 1
            continue

        if char == '"' and not field:
            in_quotes = True
            i += 1
            continue
        if char == ",":
            row.append("".join(field))
            field = []
            i += 1
            continue
        if char in ("\n", "\r"):
            row.append("".join(field))
            field = []
            if row != [""]:  # a blank line is not a record
                records.append(row)
            row = []
            i += 2 if char == "\r" and i + 1 < size and text[i + 1] == "\n" else 1
            continue
        field.append(char)
        i += 1

    if in_quotes:
        raise PhonebookFault("malformed_csv", "file ends inside a quoted field")
    if field or row:
        row.append("".join(field))
        if row != [""]:
            records.append(row)
    return records


def _csv_line(cells: list) -> str:
    return ",".join(_csv_cell(cell) for cell in cells)


def _csv_cell(cell: str) -> str:
    if any(char in QUOTE_NEEDED for char in cell):
        return '"' + cell.replace('"', '""') + '"'
    return cell

File: runtime/rust/phonebook_rt/src/io_.rs
//! Area 500 — the only block that touches the filesystem.
//!
//! The CSV state machine below is a line-for-line counterpart of the one in
//! `phonebook_rt/io_.py`. Delegating to a crate would have been shorter and
//! would have made the two backends agree only by luck; writing the same
//! explicit machine twice is what makes them agree by construction.

use std::collections::BTreeMap;

const QUOTE_NEEDED: [char; 4] = [',', '"', '\r', '\n'];

/// 500-0000001 READ_TEXT_FILE — UTF-8 strict, no newline translation.
pub fn read_text_file(path: &str) -> String {
    let bytes = match std::fs::read(path) {
        Ok(bytes) => bytes,
        Err(err) => match err.kind() {
            std::io::ErrorKind::NotFound => crate::fault("file_not_found", path),
            std::io::ErrorKind::PermissionDenied => crate::fault("permission_denied", path),
            _ => crate::fault("file_not_found", path),
        },
    };
    match String::from_utf8(bytes) {
        Ok(text) => text,
        Err(_) => crate::fault("decode_error", &format!("{path} is not valid UTF-8")),
    }
}

/// 500-0000002 WRITE_TEXT_FILE — truncating, byte-exact, no newline translation.
pub fn write_text_file(path: &str, content: &str) {
    match std::fs::write(path, content.as_bytes()) {
        Ok(()) => {}
        Err(err) => match err.kind() {
            std::io::ErrorKind::PermissionDenied => crate::fault("permission_denied", path),
            _ => crate::fault("path_not_writable", path),
        },
    }
}

/// 500-0000003 READ_CSV — header row, everything stays text.
pub fn read_csv(path: &str) -> Vec<BTreeMap<String, String>> {
    let text = read_text_file(path);
    let records = parse_csv(&text, path);
    let mut rows: Vec<BTreeMap<String, String>> = Vec::new();
    let header = match records.first() {
        Some(header) => header.clone(),
        None => return rows,
    };
    for (offset, cells) in records.iter().skip(1).enumerate() {
        if cells.len() > header.len() {
            crate::fault(
                "malformed_csv",
                &format!(
                    "{}:{} has {} cells but the header has {}",
                    path,
                    offset + 2,
                    cells.len(),
                    header.len()
                ),
            );
        }
        let mut row: BTreeMap<String, String> = BTreeMap::new();
        for (index, name) in header.iter().enumerate() {
            let value = cells.get(index).cloned().unwrap_or_default();
            row.insert(name.clone(), value);
        }
        rows.push(row);
    }
    rows
}

/// 500-0000004 WRITE_CSV — minimal quoting, LF line endings.
pub fn write_csv(path: &str, rows: &[BTreeMap<String, String>], columns: &[String]) {
    let mut out = String::new();
    out.push_str(&csv_line(columns));
    out.push('\n');
    for row in rows {
        let cells: Vec<String> = columns
            .iter()
            .map(|column| row.get(column).cloned().unwrap_or_default())
            .collect();
        out.push_str(&csv_line(&cells));
        out.push('\n');
    }
    write_text_file(path, &out);
}

// -------------------------------------------------------------------------
// the shared CSV state machine
// -------------------------------------------------------------------------

fn parse_csv(text: &str, path: &str) -> Vec<Vec<String>> {
    let chars: Vec<char> = text.chars().collect();
    let mut records: Vec<Vec<String>> = Vec::new();
    let mut row: Vec<String> = Vec::new();
    let mut field = String::new();
    let mut in_quotes = false;
    let mut i = 0usize;

    while i < chars.len() {
        let ch = chars[i];
        if in_quotes {
            if ch == '"' {
                if i + 1 < chars.len() && chars[i + 1] == '"' {
                    field.push('"');
                    i += 2;
                    continue;
                }
                in_quotes = false;
                i += 1;
                continue;
            }
            field.push(ch);
            i += 1;
            continue;
        }

        if ch == '"' && field.is_empty() {
            in_quotes = true;
            i += 1;
            continue;
        }
        if ch == ',' {
            row.push(std::mem::take(&mut field));
            i += 1;
            continue;
        }
        if ch == '\n' || ch == '\r' {
            row.push(std::mem::take(&mut field));
            if !is_blank_record(&row) {
                records.push(std::mem::take(&mut row));
            } else {
                row.clear();
            }
            i += if ch == '\r' && i + 1 < chars.len() && chars[i + 1] == '\n' {
                2
            } else {
                1
            };
            continue;
        }
        field.push(ch);
        i += 1;
    }

    if in_quotes {
        crate::fault(
            "malformed_csv",
            &format!("{path} ends inside a quoted field"),
        );
    }
    if !field.is_empty() || !row.is_empty() {
        row.push(field);
        if !is_blank_record(&row) {
            records.push(row);
        }
    }
    records
}

fn is_blank_record(row: &[String]) -> bool {
    row.len() == 1 && row[0].is_empty()
}

fn csv_line(cells: &[String]) -> String {
    let rendered: Vec<String> = cells.iter().map(|cell| csv_cell(cell)).collect();
    rendered.join(",")
}

fn csv_cell(cell: &str) -> String {
    if cell.chars().any(|c| QUOTE_NEEDED.contains(&c)) {
        format!("\"{}\"", cell.replace('"', "\"\""))
    } else {
        cell.to_string()
    }
}
